### ISSUE TITLE
Flexible debug by adding inline custom_templates

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1301]
In GitLab by @dbeltrankyl on May 23, 2024, 09:03

This issue is to add a customizable template for each job.

This is mainly useful for debugging, although it can be used to write short scripts.

The idea of adding is that instead of using PROJECT_TYPE: None, in which the default code is sleep 5 for each template, use this to have different sleep times, force jobs to fail, etc, or even only "skip" some workflow jobs. 

Usage:

```yaml
JOBS:
 JOB:
   FILE: bla.sh
   TYPE: bash
   CUSTOM_TEMPLATE: "sleep 5" 
 PYTHON_JOB:
   FILE: bla.sh
   TYPE: python3
   CUSTOM_TEMPLATE: |
      from time import sleep
      sleep(10)

```

The goal is to have a faster debug process

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 23, 2024, 09:22

Tested sucesfully, 

```yaml
JOBS:
  NOISE:
    FILE: NOISE.sh
    RUNNING: chunk
    WALLCLOCK: 00:05
    TYPE: "python3"
    CUSTOM_TEMPLATE: |
      from time import sleep
      sleep(10)
      if '%MEMBER%' == 'fc0':
          # make this fail
          return 1
      else:
          print('Hello World!')
```

Results in 

```python
#!/usr/bin/env python3

###############################################################################
#                   NOISE a053 EXPERIMENT
###############################################################################
#
#SBATCH --qos=gp_debug
#
#SBATCH -A bsc32
#
#SBATCH --cpus-per-task=1
#
#
#SBATCH -n 1
#
#SBATCH -t 00:05:00
#SBATCH -J a053_20000101_fc0_1_NOISE
#SBATCH --output=/gpfs/scratch/bsc32/bsc032070/a053/LOG_a053/a053_20000101_fc0_1_NOISE.cmd.out.0
#SBATCH --error=/gpfs/scratch/bsc32/bsc032070/a053/LOG_a053/a053_20000101_fc0_1_NOISE.cmd.err.0

#
#
###############################################################################
###################
# Autosubmit header
###################
import locale
import time
try:
    try:
        locale.setlocale(locale.LC_ALL,'C.utf8')
    except Exception as e:
        try:
            locale.setlocale(locale.LC_ALL, 'C.UTF-8')
        except Exception as e:
            try:
                locale.setlocale(locale.LC_ALL, 'en_GB')
            except Exception as e:
                locale.setlocale(locale.LC_ALL, 'es_ES')
except Exception as e:
    locale.setlocale(locale.LC_ALL, 'C')
job_name_ptrn = '/gpfs/scratch/bsc32/bsc032070/a053/LOG_a053/a053_20000101_fc0_1_NOISE'
stat_file = open(job_name_ptrn + '_STAT', 'w')
stat_file.write('int({0:.0f})\n'.format(time.time()))
stat_file.close()
###################
# Autosubmit Checkpoint
###################
# Creates a new checkpoint file upton call based on the current numbers of calls to the function
AS_CHECKPOINT_CALLS = 0
def as_checkpoint():
    global AS_CHECKPOINT_CALLS
    global job_name_ptrn
    AS_CHECKPOINT_CALLS = AS_CHECKPOINT_CALLS + 1
    open(job_name_ptrn + '_CHECKPOINT_' + str(AS_CHECKPOINT_CALLS), 'w').close()      
          
###################
# Autosubmit job
###################


from time import sleep
sleep(10)
if 'fc0' == 'fc0':
    # make this fail
    exit(1)
else:
    print('Hello World!')

###################
# Autosubmit tailer
###################

stat_file = open(job_name_ptrn + '_STAT', 'a')
stat_file.write('int({0:.0f})\n'.format(time.time()))
stat_file.close()
open(job_name_ptrn + '_COMPLETED', 'a').close()
exit(0)
```

**Comment by kinow** [comment]
From today's meeting:

- Maybe give the user some feedback if they specify both
- We need to add that to the documentation, with a note/warn which is used first, etc.
- Maybe use `script` (ecflow/cylc/...)

Great stuff! I'm going to use it!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 23, 2024, 11:09

* Added a warning indicating that FILE won't be read if defined
* Changed CUSTOM_TEMPLATE to SCRIPT:
* Changed as_conf check configuration function to, in case that File is not added, see if SCRIPT: is.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 11, 2024, 11:33

Implemented in 4.1.7
